lib/Sema/CSFix.cpp

```diff
@@ -2599,7 +2599,15 @@ bool AllowConcreteTypeSpecialization::diagnose(const Solution &solution,
 
 AllowConcreteTypeSpecialization *
 AllowConcreteTypeSpecialization::create(ConstraintSystem &cs, Type concreteTy,
+                                        ValueDecl *decl,
                                         ConstraintLocator *locator) {
+  // Check the decl and null the type to signal that the actual problem is trying to specialize a generic function
+  if (isa<AbstractFunctionDecl>(decl)) {
+    auto genericContext = decl->getAsGenericContext();
+    if (genericContext && genericContext->getGenericParams()) {
+      concreteTy = nullptr;
```

Comment:
Hm, I think this is actually a regression over what you had before but I think we need here is two fixes and two district diagnostics: one for a concrete type and one for a generic function. The latter doesn't have to carry the type but could benefit from having a declaration since we could add a note that points out where the function/method is declared.

Reply:
Yeah, I should've made it a separate fix type the whole time, and these were bad shortcuts.